File: client-sdks/go/sdk/comm/message_codec.go
// ------------------------------------------------------------------------------
// Copyright 2025 to Present, Jason Lam - VertexCache
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ------------------------------------------------------------------------------

package comm

import (
	"bytes"
	"encoding/binary"
	"errors"
	"fmt"
	"io"
)

const (
	MaxMessageSize  = 10 * 1024 * 1024 // 10MB
	ProtocolVersion = 0x00000101
)

// WriteFramedMessage writes a framed message:
// [4 bytes length][4 bytes protocol version][payload]
func WriteFramedMessage(w io.Writer, payload []byte) error {
	if len(payload) > MaxMessageSize {
		return fmt.Errorf("message too large: %d", len(payload))
	}

	var header bytes.Buffer

	// Write length (int32, big-endian)
	if err := binary.Write(&header, binary.BigEndian, int32(len(payload))); err != nil {
		return err
	}

	// Write protocol version (uint32, big-endian)
	if err := binary.Write(&header, binary.BigEndian, uint32(ProtocolVersion)); err != nil {
		return err
	}

	// Write header + payload
	if _, err := w.Write(header.Bytes()); err != nil {
		return err
	}
	if _, err := w.Write(payload); err != nil {
		return err
	}
	return nil
}

// ReadFramedMessage reads a framed message:
// [4 bytes length][4 bytes protocol version][payload]
func ReadFramedMessage(r io.Reader) ([]byte, error) {
	header := make([]byte, 8)
	n, err := io.ReadFull(r, header)
	if err != nil {
		if err == io.ErrUnexpectedEOF || err == io.EOF || n < 8 {
			return nil, nil // treat short header as non-fatal
		}
		return nil, err
	}

	length := int32(binary.BigEndian.Uint32(header[0:4]))
	version := binary.BigEndian.Uint32(header[4:8])

	if version != ProtocolVersion {
		return nil, errors.New("unsupported protocol version")
	}
	if length <= 0 || length > MaxMessageSize {
		return nil, fmt.Errorf("invalid message length: %d", length)
	}

	payload := make([]byte, length)
	if _, err := io.ReadFull(r, payload); err != nil {
		return nil, err
	}
	return payload, nil
}

File: client-sdks/python/sdk/comm/message_codec.py
# ------------------------------------------------------------------------------
# Copyright 2025 to Present, Jason Lam - VertexCache (https://github.com/vertexcache/vertexcache)
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ------------------------------------------------------------------------------

import struct

MAX_MESSAGE_SIZE = 10 * 1024 * 1024  # 10MB
PROTOCOL_VERSION = 0x01


def write_framed_message(payload: bytes) -> bytes:
    """
    Writes a framed message using the VertexCache protocol format.

    Format:
    - 4 bytes big-endian integer representing the payload length
    - 1 byte protocol version
    - N bytes payload data

    Args:
        payload (bytes): The message payload.

    Returns:
        bytes: The framed binary message.

    Raises:
        TypeError: If payload is not a bytes object.
        ValueError: If the payload exceeds the maximum allowed size.
    """
    if not isinstance(payload, bytes):
        raise TypeError("Payload must be bytes")

    if len(payload) > MAX_MESSAGE_SIZE:
        raise ValueError(f"Message too large: {len(payload)}")

    length_bytes = struct.pack(">I", len(payload))
    version_byte = struct.pack("B", PROTOCOL_VERSION)
    return length_bytes + version_byte + payload


def read_framed_message(buffer: bytes):
    """
    Reads a framed message from a byte buffer using the VertexCache protocol.

    Format:
    - 4 bytes big-endian integer representing the payload length
    - 1 byte protocol version
    - N bytes payload data

    Args:
        buffer (bytes): The raw byte buffer to decode.

    Returns:
        tuple[bytes, bytes] | None:
            - A tuple (payload, remaining) if a complete message is decoded
            - None if the buffer is too short to decode the header or full payload

    Raises:
        ValueError: If the protocol version is unsupported or the message length is invalid.
    """
    if len(buffer) < 5:
        return None

    length, version = struct.unpack(">IB", buffer[:5])

    if version != PROTOCOL_VERSION:
        raise ValueError(f"Unsupported protocol version: {version}")

    if length <= 0 or length > MAX_MESSAGE_SIZE:
        raise ValueError(f"Invalid message length: {length}")

    if len(buffer) < 5 + length:
        return None

    payload = buffer[5:5+length]
    remaining = buffer[5+length:]
    return payload, remaining
